benches/hashmap.rs

```diff
@@ -1,285 +1,445 @@
+    fn insert_clone(&self, k: K, v: V) -> Self;
+    fn remove(&mut self, k: &K) -> Option<V>;
+    fn remove_clone(&self, k: &K) -> Self;
+    fn get<Q>(&self, k: &Q) -> Option<&V>
+    where
+        K: Borrow<Q>,
+        Q: Hash + Eq + ?Sized;
+    fn iter(&self) -> Self::Iter<'_>;
+}
+
+// Implementation for imbl::HashMap
+impl<K, V> BenchMap<K, V> for HashMap<K, V>
+where
+    K: Clone + Hash + Eq,
+    V: Clone,
+{
+    type Iter<'a>
+        = imbl::hashmap::Iter<'a, K, V, imbl::shared_ptr::DefaultSharedPtr>
+    where
+        K: 'a,
+        V: 'a;
+
+    fn new() -> Self {
+        HashMap::new()
+    }
 
-#![feature(test)]
+    fn insert(&mut self, k: K, v: V) -> Option<V> {
+        self.insert(k, v)
+    }
 
-extern crate imbl;
-extern crate rand;
-extern crate test;
+    fn insert_clone(&self, k: K, v: V) -> Self {
+        self.update(k, v)
+    }
 
-use rand::{rngs::SmallRng, Rng, SeedableRng};
-use std::iter::FromIterator;
-use test::{black_box, Bencher};
+    fn remove(&mut self, k: &K) -> Option<V> {
+        self.remove(k)
+    }
 
-use imbl::hashmap::HashMap;
+    fn remove_clone(&self, k: &K) -> Self {
+        self.without(k)
+    }
 
-fn random_keys(size: usize) -> Vec<i64> {
-    let mut gen = SmallRng::seed_from_u64(1);
-    let mut set = Vec::new();
-    while set.len() < size {
-        let next = gen.random::<i64>();
-        if !set.contains(&next) {
-            set.push(next);
-        }
+    fn get<Q>(&self, k: &Q) -> Option<&V>
+    where
+        K: Borrow<Q>,
+        Q: Hash + Eq + ?Sized,
+    {
+        self.get(k)
     }
-    set
-}
 
-fn reorder<A: Copy>(vec: &[A]) -> Vec<A> {
-    let mut gen = SmallRng::seed_from_u64(1);
-    let mut set = vec.to_owned();
-    let mut out = Vec::new();
-    while !set.is_empty() {
-        let i = gen.random::<u64>() as usize % set.len();
-        let v = set.remove(i);
-        out.push(v)
+    fn iter(&self) -> Self::Iter<'_> {
+        self.iter()
     }
-    out
 }
 
-fn hashmap_lookup_n(size: usize, b: &mut Bencher) {
-    let keys = random_keys(size);
-    let order = reorder(&keys);
-    let m: HashMap<i64, i64> = HashMap::from_iter(keys.into_iter().map(|i| (i, 1)));
-    b.iter(|| {
-        for i in &order {
-            let _ = m.get(i);
-        }
-    })
-}
+// Implementation for std::collections::HashMap
+impl<K, V> BenchMap<K, V> for StdHashMap<K, V>
+where
+    K: Clone + Hash + Eq,
+    V: Clone,
+{
+    const IMMUTABLE: bool = false;
+    type Iter<'a>
+        = std::collections::hash_map::Iter<'a, K, V>
+    where
+        K: 'a,
+        V: 'a;
+
+    fn new() -> Self {
+        StdHashMap::new()
+    }
 
-#[bench]
-fn hashmap_lookup_10(b: &mut Bencher) {
-    hashmap_lookup_n(10, b)
-}
+    fn insert(&mut self, k: K, v: V) -> Option<V> {
+        self.insert(k, v)
+    }
 
-#[bench]
-fn hashmap_lookup_100(b: &mut Bencher) {
-    hashmap_lookup_n(100, b)
-}
+    fn insert_clone(&self, _k: K, _v: V) -> Self {
+        Self::new()
```

Comment:
I'd be inclined to just implement this as `let mut ret = self.clone(); ret.insert(k, v); ret`, even though you never actually benchmark it. Right now, this PR is skipping mutable collections in three places: here, in `bench_insert/bench_remove`, and then again in `bench_group`. I'd prefer just to skip them in one place (probably `bench_group`), because then if someone wants to actually run those benchmarks -- for example, to demonstrate a situation where immutable collections are important -- then it will be less confusing for them.

Reply:
Makes sense, done in 919dcd5